File: ED del diego/Haskell/Julio 2022/haskell/WBinTree.hs
-------------------------------------------------------------------------------
-- Student's name:
-- Student's group:
-- Identity number (DNI if Spanish/passport if Erasmus):
--
-- Data Structures. Grado en Informática. UMA.
-------------------------------------------------------------------------------

module WBinTree( WBinTree
               , empty
               , insert
               , isWeightBalanced
               , mkWBinTree
               ) where

data WBinTree a = Empty
                | Node Int a (WBinTree a) (WBinTree a)
                deriving Show

-- ESCRIBE TU SOLUCIÓN DEBAJO ----------------------------------------------
-- WRITE YOUR SOLUTION BELOW  ----------------------------------------------
-- EXERCISE 4

isWeightBalanced :: WBinTree a -> Bool
isWeightBalanced Empty = True
isWeightBalanced (Node _ _ Empty Empty) = True
isWeightBalanced (Node _ _ lh Empty) = True
isWeightBalanced (Node _ _ Empty rh) = True
isWeightBalanced (Node _ _ (Node wL _ ll rl) (Node wR _ lr rr)) = wL - wR < 1

insert :: a -> WBinTree a -> WBinTree a
insert x Empty = Node 1 x Empty Empty
insert x (Node w y Empty rh) = Node (w+1) y (insert x Empty) rh
insert x (Node w y lh Empty) = Node (w+1) y lh (insert x Empty)
insert x wb@(Node w y lh rh)
    | isWeightBalanced wb = Node (w+1) y (insert x lh) rh
    | otherwise = Node (w+1) y lh (insert x rh)



-- | NO MODIFICAR A PARTIR DE AQUÍ --------------------------------------------
-- | DO NOT MODIFY CODE BELOW      --------------------------------------------

empty :: WBinTree a
empty = Empty

mkWBinTree :: [a] -> WBinTree a
mkWBinTree = foldr insert empty

File: ED del diego/Java/Julio 2022/src/exercises/WBinTree.java
/* ------------------------------------------------------------------------------
   -- Student's name:
   -- Student's group:
   -- Identity number (DNI if Spanish/passport if Erasmus):
   --
   -- Data Structures. Grado en Informática. UMA.
   -------------------------------------------------------------------------------
*/

package exercises;

public class WBinTree<T> {

    private static class Node<E> {
        private E item;
        private int weight;
        private Node<E> left, right;

        Node(E x, int w, Node<E> lt, Node<E> rt) {
            item= x;
            weight = w;
            left= lt;
            right = rt;
        }
    }

    private Node<T> root;

    public WBinTree() {
        root = null;
    }

    // -- ESCRIBE TU SOLUCIÓN DEBAJO ----------------------------------------------
    // -- WRITE YOUR SOLUTION BELOW  ----------------------------------------------
    // -- EXERCISE 4

    public boolean isWeightBalanced() {
        // TODO
        if(root == null){
            return true;
        }else{
            int wL = 0;
            int wR = 0;
            if(root.left != null){
                wL = root.left.weight;
            }
            if(root.right != null){
                wR = root.right.weight;
            }
            return wL - wR < 1;
        }
    }

    public void insert(T x) {
        // TODO
        // Si el nodo es vacío
        root = insertRec(x,root);
    }

    private <T> Node<T> insertRec(T x, Node<T> root){
        if(root == null){
            root = new Node<>(x,1,null,null);
        }else{
            root.weight++;
            if(root.left == null){
                root.left = insertRec(x, null);
            }else if(root.right == null){
                root.right = insertRec(x,null);
            }else{
                if(isWeightBalanced()){
                    root.left = insertRec(x,root.left);
                }else{
                    root.right = insertRec(x,root.right);
                }
            }
        }
        return root;
    }
    // -- DO NOT MODIFY CODE BELOW -----------------------------------------------------------
    // -- NO MODIFIQUES DEBAJO ---------------------------------------------------------------

    @Override
    public String toString() {
        StringBuffer sb = new StringBuffer();
        toStringRec(root, sb);
        return sb.toString();
    }

    private static <T> void toStringRec(Node<T> node, StringBuffer sb) {
        if (node == null) {
            sb.append("null");
        }
        else {
            sb.append(String.format("Node(%d, %s, ", node.weight, node.item));
            toStringRec(node.left, sb);
            sb.append(", ");
            toStringRec(node.right, sb);
            sb.append(")");
        }
    }
}
